## Data Exploration & Loading ##


**Extract and Cleansing Data**

In [62]:
# import libraries
import pandas as pd
import psycopg2
from sqlalchemy import create_engine

In [63]:
# Load dataset
df = pd.read_csv('Walmart.csv', encoding_errors='ignore')

In [64]:
# Dataset general information
df.head(10)


invoice_id   Branch         City                category unit_price  \
0           1  WALM003  San Antonio       Health and beauty     $74.69   
1           2  WALM048    Harlingen  Electronic accessories     $15.28   
2           3  WALM067  Haltom City      Home and lifestyle     $46.33   
3           4  WALM064      Bedford       Health and beauty     $58.22   
4           5  WALM013       Irving       Sports and travel     $86.31   
5           6  WALM026       Denton  Electronic accessories     $85.39   
6           7  WALM088     Cleburne  Electronic accessories     $68.84   
7           8  WALM100       Canyon      Home and lifestyle     $73.56   
8           9  WALM066    Grapevine       Health and beauty     $36.26   
9          10  WALM065   Texas City      Food and beverages     $54.84   

   quantity      date      time payment_method  rating  profit_margin  
0       7.0  05/01/19  13:08:00        Ewallet     9.1           0.48  
1       5.0  08/03/19  10:29:00           Cash     9.6           0.48  
2       7.0  03/03/19  13:23:00    Credit card     7.4           0.33  
3       8.0  27/01/19  20:33:00        Ewallet     8.4           0.33  
4       7.0  08/02/19  10:37:00        Ewallet     5.3           0.48  
5       7.0  25/03/19  18:30:00        Ewallet     4.1           0.48  
6       6.0  25/02/19  14:36:00        Ewallet     5.8           0.33  
7      10.0  24/02/19  11:38:00        Ewallet     8.0           0.18  
8       2.0  10/01/19  17:15:00    Credit card     7.2           0.33  
9       3.0  20/02/19  13:27:00    Credit card     5.9           0.33

In [65]:
df.describe()

invoice_id      quantity        rating  profit_margin
count  10051.000000  10020.000000  10051.000000   10051.000000
mean    5025.741220      2.353493      5.825659       0.393791
std     2901.174372      1.602658      1.763991       0.090669
min        1.000000      1.000000      3.000000       0.180000
25%     2513.500000      1.000000      4.000000       0.330000
50%     5026.000000      2.000000      6.000000       0.330000
75%     7538.500000      3.000000      7.000000       0.480000
max    10000.000000     10.000000     10.000000       0.570000

In [66]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10051 entries, 0 to 10050
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   invoice_id      10051 non-null  int64  
 1   Branch          10051 non-null  object 
 2   City            10051 non-null  object 
 3   category        10051 non-null  object 
 4   unit_price      10020 non-null  object 
 5   quantity        10020 non-null  float64
 6   date            10051 non-null  object 
 7   time            10051 non-null  object 
 8   payment_method  10051 non-null  object 
 9   rating          10051 non-null  float64
 10  profit_margin   10051 non-null  float64
dtypes: float64(3), int64(1), object(7)
memory usage: 863.9+ KB


This dataset contains 6435 rows(observations) and 8 columns(attributes). In unit_price, there is '$' which is very hard to do any calculation needed and it is classified as 'object', which is not correct. There also 10020 record, indicating that there might some missing data.

In [67]:
# checking duplicates
df[df.duplicated(keep=False)].sort_values(by='invoice_id')

invoice_id   Branch        City             category unit_price  \
9949         9950  WALM038  Sugar Land  Fashion accessories        $17   
10000        9950  WALM038  Sugar Land  Fashion accessories        $17   
10001        9951  WALM082     Weslaco   Home and lifestyle        $58   
9950         9951  WALM082     Weslaco   Home and lifestyle        $58   
9951         9952  WALM035  San Angelo  Fashion accessories        $76   
...           ...      ...         ...                  ...        ...   
10048        9998  WALM050    Victoria  Fashion accessories        $52   
10049        9999  WALM032       Tyler   Home and lifestyle        $79   
9998         9999  WALM032       Tyler   Home and lifestyle        $79   
9999        10000  WALM069    Rockwall  Fashion accessories        $62   
10050       10000  WALM069    Rockwall  Fashion accessories        $62   

       quantity      date      time payment_method  rating  profit_margin  
9949        1.0  27/11/23   9:15:00           Cash     3.0           0.48  
10000       1.0  27/11/23   9:15:00           Cash     3.0           0.48  
10001       2.0  08/07/20  12:39:00           Cash     6.0           0.33  
9950        2.0  08/07/20  12:39:00           Cash     6.0           0.33  
9951        3.0  02/10/21  16:34:00           Cash     6.0           0.48  
...         ...       ...       ...            ...     ...            ...  
10048       3.0  15/06/23  16:00:00    Credit card     4.0           0.48  
10049       2.0  25/02/21  12:25:00           Cash     7.0           0.48  
9998        2.0  25/02/21  12:25:00           Cash     7.0           0.48  
9999        3.0  26/09/20   9:48:00           Cash     3.0           0.33  
10050       3.0  26/09/20   9:48:00           Cash     3.0           0.33  

[102 rows x 11 columns]

In [68]:
# checking if there missing value
df.isnull().sum() ## check total missing value by column

invoice_id         0
Branch             0
City               0
category           0
unit_price        31
quantity          31
date               0
time               0
payment_method     0
rating             0
profit_margin      0
dtype: int64

In [69]:
## inspect missing rows
df[df.isnull().any(axis=1)]

invoice_id   Branch           City             category unit_price  \
1892        1893  WALM024     Carrollton   Home and lifestyle        NaN   
1893        1894  WALM009          Plano  Fashion accessories        NaN   
1894        1895  WALM010         Laredo   Home and lifestyle        NaN   
1895        1896  WALM069       Rockwall  Fashion accessories        NaN   
1896        1897  WALM093       Angleton   Home and lifestyle        NaN   
1897        1898  WALM098  Mineral Wells  Fashion accessories        NaN   
1898        1899  WALM009          Plano   Home and lifestyle        NaN   
1899        1900  WALM071         Lufkin  Fashion accessories        NaN   
1900        1901  WALM090      Brownwood   Home and lifestyle        NaN   
1901        1902  WALM067    Haltom City  Fashion accessories        NaN   
1902        1903  WALM033       Pearland   Home and lifestyle        NaN   
1903        1904  WALM018         Frisco  Fashion accessories        NaN   
1904        1905  WALM076     Huntsville   Home and lifestyle        NaN   
1905        1906  WALM010         Laredo  Fashion accessories        NaN   
1906        1907  WALM030     Richardson   Home and lifestyle        NaN   
1907        1908  WALM001        Houston  Fashion accessories        NaN   
1908        1909  WALM096     Eagle Pass   Home and lifestyle        NaN   
1909        1910  WALM085      Kerrville  Fashion accessories        NaN   
1910        1911  WALM040       Edinburg   Home and lifestyle        NaN   
1911        1912  WALM080    Nacogdoches  Fashion accessories        NaN   
1912        1913  WALM068       Burleson   Home and lifestyle        NaN   
1913        1914  WALM048      Harlingen  Fashion accessories        NaN   
1914        1915  WALM018         Frisco   Home and lifestyle        NaN   
1915        1916  WALM015  Grand Prairie  Fashion accessories        NaN   
1916        1917  WALM041        Mission   Home and lifestyle        NaN   
1917        1918  WALM045  Missouri City  Fashion accessories        NaN   
1918        1919  WALM011        Lubbock   Home and lifestyle        NaN   
1919        1920  WALM052      Mansfield  Fashion accessories        NaN   
1920        1921  WALM022       Mesquite   Home and lifestyle        NaN   
1921        1922  WALM024     Carrollton  Fashion accessories        NaN   
1922        1923  WALM070          Hurst   Home and lifestyle        NaN   

      quantity      date      time payment_method  rating  profit_margin  
1892       NaN  09/06/23  16:09:00        Ewallet     9.0           0.48  
1893       NaN  02/08/22  22:11:00        Ewallet     4.0           0.48  
1894       NaN  01/08/23  12:06:00        Ewallet     4.0           0.48  
1895       NaN  21/07/23  15:49:00        Ewallet     4.0           0.33  
1896       NaN  15/05/22  12:36:00        Ewallet     9.0           0.33  
1897       NaN  01/11/23  10:08:00        Ewallet     4.0           0.18  
1898       NaN  01/10/21  13:05:00        Ewallet     6.0           0.48  
1899       NaN  19/05/21   8:36:00        Ewallet     7.0           0.33  
1900       NaN  13/02/21  14:50:00        Ewallet     8.0           0.33  
1901       NaN  15/05/22  19:20:00        Ewallet     8.0           0.33  
1902       NaN  28/08/22  15:37:00        Ewallet     4.0           0.48  
1903       NaN  21/04/22  18:30:00        Ewallet     4.0           0.48  
1904       NaN  02/03/20  14:33:00        Ewallet     4.0           0.33  
1905       NaN  20/08/21  14:25:00        Ewallet     5.0           0.48  
1906       NaN  14/12/23  18:12:00        Ewallet     8.0           0.48  
1907       NaN  09/08/22  12:46:00        Ewallet     9.0           0.36  
1908       NaN  23/11/21  18:26:00        Ewallet     9.0           0.33  
1909       NaN  24/08/22  11:53:00        Ewallet     4.0           0.33  
1910       NaN  27/07/20  21:38:00        Ewallet     9.0           0.48  
1911       NaN  15/10/20   8:01:00        Ewallet     7.0           0.33 

There are 51 duplicates, and it completely duplicated from previous one, this might be the problems in data collecting, so remove them the best choice. For missing value, 'unit_price' and 'quantity' both have 31 missing value. In detail, missing row basically just missing information of that row and is not observed any further hidden problems. I prefer to drop all the rows that contain missing value because in this such big dataset, 31 rows won't affect significantly to the overall perspective.

In [70]:
# Treatment
df.drop_duplicates(inplace=True)

In [71]:
df.dropna(inplace=True)

In [72]:
# check the new df 
df.shape

(9969, 11)

After removing all the duplicates and rows that contain null value, the dataframe now has 9969 rows(observations).

In [73]:
# convert 'unit_price' from to float
## remove $
df['unit_price'] = df['unit_price'].str.replace('$','').astype(float)

df.head()

invoice_id   Branch         City                category  unit_price  \
0           1  WALM003  San Antonio       Health and beauty       74.69   
1           2  WALM048    Harlingen  Electronic accessories       15.28   
2           3  WALM067  Haltom City      Home and lifestyle       46.33   
3           4  WALM064      Bedford       Health and beauty       58.22   
4           5  WALM013       Irving       Sports and travel       86.31   

   quantity      date      time payment_method  rating  profit_margin  
0       7.0  05/01/19  13:08:00        Ewallet     9.1           0.48  
1       5.0  08/03/19  10:29:00           Cash     9.6           0.48  
2       7.0  03/03/19  13:23:00    Credit card     7.4           0.33  
3       8.0  27/01/19  20:33:00        Ewallet     8.4           0.33  
4       7.0  08/02/19  10:37:00        Ewallet     5.3           0.48

In [74]:
# convert day, time to correct format
df['date'] = pd.to_datetime(df['date'], format='%d/%m/%y')

In [75]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9969 entries, 0 to 9999
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   invoice_id      9969 non-null   int64         
 1   Branch          9969 non-null   object        
 2   City            9969 non-null   object        
 3   category        9969 non-null   object        
 4   unit_price      9969 non-null   float64       
 5   quantity        9969 non-null   float64       
 6   date            9969 non-null   datetime64[ns]
 7   time            9969 non-null   object        
 8   payment_method  9969 non-null   object        
 9   rating          9969 non-null   float64       
 10  profit_margin   9969 non-null   float64       
dtypes: datetime64[ns](1), float64(4), int64(1), object(5)
memory usage: 934.6+ KB


After change datatype for 'unit_price', it is now numerical.

In [81]:
# Creating new columns called 'Total_sales' and ''Total_profit', which is the total revenue of each invoice id
df['total_sales'] = df['unit_price'] * df['quantity']

df['total_profit'] = df['total_sales'] * df['profit_margin']

In [82]:
# check the cleaned dataset
df.head()

invoice_id   branch         city                category  unit_price  \
0           1  WALM003  San Antonio       Health and beauty       74.69   
1           2  WALM048    Harlingen  Electronic accessories       15.28   
2           3  WALM067  Haltom City      Home and lifestyle       46.33   
3           4  WALM064      Bedford       Health and beauty       58.22   
4           5  WALM013       Irving       Sports and travel       86.31   

   quantity       date      time payment_method  rating  profit_margin  \
0       7.0 2019-01-05  13:08:00        Ewallet     9.1           0.48   
1       5.0 2019-03-08  10:29:00           Cash     9.6           0.48   
2       7.0 2019-03-03  13:23:00    Credit card     7.4           0.33   
3       8.0 2019-01-27  20:33:00        Ewallet     8.4           0.33   
4       7.0 2019-02-08  10:37:00        Ewallet     5.3           0.48   

    total  total_sales  total_profit  
0  522.83       522.83      250.9584  
1   76.40        76.40       36.6720  
2  324.31       324.31      107.0223  
3  465.76       465.76      153.7008  
4  604.17       604.17      290.0016

**Loading to Parquet**

In [84]:
df.to_parquet('cleaned_walamrt_data.parquet')

**Loading to SQL**

In [78]:
# psql connection
## create a connection
conn = psycopg2.connect(
    host='localhost', # <-- This forces TCP/IP
    database='walmart_db',
    user='postgres',
    password='Thuan18072004@'
)
### because the password contain special character, so need to use URL-encoded, here '@' = '%40'
### if there any special char, need to change it by using URL encoded in 'create_engine'
engine_sql = create_engine("postgresql+psycopg2://postgres:Thuan18072004%40@localhost:5432/walmart_db")

try:
    engine_sql
    print("Connection Seccessed to PSQL")
except:
    print("Unable to connect")

Connection Seccessed to PSQL


In [79]:
## before using psql for performing task, make sure all the cols name are in lowercase, otherwise it will harder to call

## convert all cols' name to lowercase
df.columns=df.columns.str.lower()

In [80]:



# connect to psql
df.to_sql(name='walmart', con=engine_sql, if_exists='replace', index=False)

## note: make sure when name a table, if using uppercase, use quote when calling table name because psql will convert to lowercase and search for table
## the double quotes tell PostgreSQL to look for the table with the exact case you typed, 
## Without quotes, PostgreSQL automatically converts the name to all lowercase and looks for a table 

969